ipfix/marshal.go

```diff
@@ -54,25 +54,44 @@ func (m *Message) JSONMarshal(b *bytes.Buffer, datasetIndex int) ([]byte, error)
 
 func (m *Message) encodeDataSet(b *bytes.Buffer, i int) error {
 	var (
-		length   int
+		num_fields   int
+		num_repeats int
+		counter int
 		err      error
 	)
 
-	length = len(m.DataSets[i])
+	num_fields = len(m.DataSets[i])
+	counter = 0
 	b.WriteString("\"Data\":{")
-	for j := range m.DataSets[i] {
+	for eKey, fields := range m.DataSets[i] {
+		num_repeats = len(fields)
+		counter++
+
 		b.WriteByte('"')
-		b.WriteString(strconv.FormatInt(int64(m.DataSets[i][j].EnterpriseNo), 10))
+		b.WriteString(strconv.FormatInt(int64(eKey.EnterpriseNo), 10))
 		b.WriteByte('_')
-		b.WriteString(strconv.FormatInt(int64(m.DataSets[i][j].ID), 10))
+		b.WriteString(strconv.FormatInt(int64(eKey.ElementID), 10))
 		b.WriteString("\":")
-		err = m.writeValue(b, i, j)
 
-		if j < length-1 {
-			b.WriteString(",")
+		if num_repeats == 1 {
+			err = m.writeValue(b, eKey, i, 0)
+		} else {
+			b.WriteByte('[')
+			for j := range fields {
+				err = m.writeValue(b, eKey, i, j)
+				if j < num_repeats - 1 {
+					b.WriteByte(',')
+				}
+			}
+			b.WriteByte(']')
+		}
```

Comment:
I'm concerned about this logic. For a field that can have `>= 1` repeats, the column could switch between being a JSON array or not. Does this impact parsing at all?

Reply:
We just confirmed that STS handles this with the `to_array("key")` function, so we should be good on that front.